check for x version updates

DebugSteven opened the pull request:
This PR adds a check to tidy to assert that the installed version of `x` is equal to the version in `src/tools/x/Cargo.toml`. It checks the installed version of `x` by parsing the output of `cargo install --list` (as an option proposed in this [issue](https://github.com/rust-lang/rust/issues/106469)).

It does not warn if `x` has not yet been installed, on the assumption that the user isn't interested in using it.

DebugSteven committed:
run cargo install to check for x installation and version

albertlarsan68 commented:
The code itself works, but it is never called.
Also, you are missing the `semver` dependency.

Here is the patch that got me to get it to work: https://gist.github.com/albertlarsan68/bd298307c84efbfcc2ee2edfd50b2ba5
r? @albertlarsan68
@rustbot author

DebugSteven commented:
Hey, thank you for catching that. This commit was originally continuing off of this [branch](https://github.com/rust-lang/rust/pull/104552/files) & I forgot to pull over the rest of the changes. There was one other change I made on that branch that I moved over as well, but otherwise I did almost exactly what you wrote. I previously tested this change by running `x test tidy` & also made sure that it worked with an arbitrary `x` executable, in the output of `cargo install --list`, that didn’t have the `src/tools/x` path.

For the other change I moved over, the `check` macro has a leading comma if you don’t have any path `args`. I ran into this because at one point I didn’t need any of the paths & tried to call `check!(x_version)`. I believe it expanded to `check!(, x_version)` & error’d out because of the syntax. I’d need to reproduce it to double check. Let me know if you’d like me to squash these commits or have a separate PR for the macro change.

DebugSteven commented:
One remaining concern I have is: should any of this code be using `tidy_error`? I’m thinking about the feedback in this [issue](https://github.com/rust-lang/rust/issues/106469) that tidy shouldn’t emit an error if it sees old versions of `x`. I think it would make sense to throw an error if cargo fails to run, but would it be more appropriate to warn a user with `println` for the other errors here?

albertlarsan68 commented:
Thanks for your PR!
If you can extract the macro change into another PR and assign it to me, it would be great.
I think squashing the 2 first commits is also a good idea.
Post `@rustbot ready` when it is done.

DebugSteven commented:
@rustbot ready

bors commented:
:pushpin: Commit 540ca2ff2acf5539868b1f1c156f6c895fdc9ffc has been approved by `albertlarsan68`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=albertlarsan68 540ca2ff2acf5539868b1f1c156f6c895fdc9ffc -->
<!-- homu: {"type":"Approved","sha":"540ca2ff2acf5539868b1f1c156f6c895fdc9ffc","approver":"albertlarsan68","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

compiler-errors commented:
@albertlarsan68, for reference, [`rollup=iffy`](https://forge.rust-lang.org/compiler/reviews.html#rollups) is used when the PR has a risk of failing in CI. It doesn't affect whether or not we can revert the PR in the future.

I think this can be marked as maybe.
@bors rollup=maybe
